exp/internal/controllers/machinepool_controller_phases.go

```
@@ -197,48 +197,51 @@ func (r *MachinePoolReconciler) reconcileBootstrap(ctx context.Context, cluster
-		return ctrl.Result{RequeueAfter: externalReadyWait}, nil
+		// Get and set the name of the secret containing the bootstrap data.
+		secretName, _, err := unstructured.NestedString(bootstrapConfig.Object, "status", "dataSecretName")
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to retrieve dataSecretName from bootstrap provider for MachinePool %q in namespace %q", m.Name, m.Namespace)
+		} else if secretName == "" {
+			return ctrl.Result{}, errors.Errorf("retrieved empty dataSecretName from bootstrap provider for MachinePool %q in namespace %q", m.Name, m.Namespace)
+		}
+
+		m.Spec.Template.Spec.Bootstrap.DataSecretName = pointer.String(secretName)
+		m.Status.BootstrapReady = true
+		return ctrl.Result{}, nil
 	}
 
-	// Get and set the name of the secret containing the bootstrap data.
-	secretName, _, err := unstructured.NestedString(bootstrapConfig.Object, "status", "dataSecretName")
-	if err != nil {
-		return ctrl.Result{}, errors.Wrapf(err, "failed to retrieve dataSecretName from bootstrap provider for MachinePool %q in namespace %q", m.Name, m.Namespace)
-	} else if secretName == "" {
-		return ctrl.Result{}, errors.Errorf("retrieved empty dataSecretName from bootstrap provider for MachinePool %q in namespace %q", m.Name, m.Namespace)
+	// If dataSecretName is set without a ConfigRef, this means the user brought their own bootstrap data.
+	if m.Spec.Template.Spec.Bootstrap.DataSecretName != nil {
+		m.Status.BootstrapReady = true
+		conditions.MarkTrue(m, clusterv1.BootstrapReadyCondition)
+		return ctrl.Result{}, nil
 	}
 
-	m.Spec.Template.Spec.Bootstrap.DataSecretName = pointer.String(secretName)
-	m.Status.BootstrapReady = true
-	return ctrl.Result{}, nil
+	// Return error if neither ConfigRef nor DataSecretName are set.
+	return ctrl.Result{}, errors.Errorf("neither .spec.bootstrap.configRef nor .spec.bootstrap.dataSecretName are set for MachinePool %q in namespace %q", m.Name, m.Namespace)
```

@sbueringer Is this what you had in mind? It's not too difficult to update the `dataSecretName` when the `configRef != nil`, but this refactor introduces a bit of an edge case where I'm not sure how to proceed when both are nil. In fact I believe that should be an impossible case based on the webhook.

---

correct, this should never happen because the webhook disallows it: https://github.com/kubernetes-sigs/cluster-api/blob/main/exp/api/v1beta1/machinepool_webhook.go#L107

---

Do you think we should just return an error like this then? It seems a bit awkward since it's an impossible case. Alternatively I could try to refactor some more.

---

I think we can return an error but add a comment above that this should never happen and point to the webhook validation